src/datachain/client/gcs.py

```diff
@@ -33,13 +33,21 @@ def create_fs(cls, **kwargs) -> GCSFileSystem:
         return cast(GCSFileSystem, super().create_fs(**kwargs))
 
     def url(self, path: str, expires: int = 3600, **kwargs) -> str:
-        try:
-            return self.fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
-        except AttributeError as exc:
-            is_anon = self.fs.storage_options.get("token") == "anon"
-            if is_anon and "you need a private key to sign credentials" in str(exc):
-                return f"https://storage.googleapis.com/{self.name}/{path}"
-            raise
+        """
+        Generate a signed URL for the given path.
+        If the client is anonymous, a public URL is returned instead.
+        """
+        fs = cast(GCSFileSystem, self.fs)
+        if fs.storage_options.get("token") == "anon":
+            return self.public_url(path)
+        return fs.sign(self.get_full_path(path), expiration=expires, **kwargs)
+
+    def public_url(self, path: str) -> str:
```

Comment:
Please, let's not extend `Client` - not every `path` has a `public_url`, and it feels like it's just a special condition that could be handled by `url()` itself.

Also, do you mind opening a bug report on `gcsfs`? Maybe they will be able to suggest a better alternative?

Reply:
> Please, let's not extend `Client` - not every `path` has a `public_url`, and it feels like it's just a special condition that could be handled by `url()` itself.

I am not extending `Client` here (yet), I am adding extra method to `GCSClient` for better readability.

Also the idea in general (for future) is to have `public_url` method in client so in case we are sure we have no credentials and bucket is public we can call `public_url` right away, without running all complex logic and checks inside clients (e.g. `gcsfs`).

> Also, do you mind opening a bug report on `gcsfs`? Maybe they will be able to suggest a better alternative?

I am considering to open a PR in `gcsfs`, but this will require much more work and I am still not sure, to be honest, if this is a way to go for the `gcsfs` or this is only specific for our use case, since `fs.sign` method is about signing URL, and if it will return a public URL instead of an error it may not be something user is waiting for.

In case of AWS S3 it is how `s3` library works and `fsspec` just proxying an error, in case of GCP Storage it is how gcloud library works and `fsspec` just proxying public URL. It is kind of complicated in general case and needed much more work comparing to the changes in this PR where we do know we do need public URL for sure.

Reply:
> Please, let's not extend `Client` - not every `path` has a `public_url`, and it feels like it's just a special condition that could be handled by `url()` itself.

Updated the code to remove extra method.

Reply:
Thank you, excellent points. I am okay with adding public_url.

---

Btw, have you seen this: [GCSFileSystem.url](https://github.com/fsspec/gcsfs/blob/378cae6f56f3dbca1b909d58f21a8d61ed5b0a45/gcsfs/core.py#L1008)? Maybe we can use this instead of hardcoding url?

Reply:
> Thank you, excellent points. I am okay with adding public_url.

❤️ Will be back to this in the follow-up.

> Btw, have you seen this: [GCSFileSystem.url](https://github.com/fsspec/gcsfs/blob/378cae6f56f3dbca1b909d58f21a8d61ed5b0a45/gcsfs/core.py#L1008)? Maybe we can use this instead of hardcoding url?

I saw it but I am not sure what kind of an URL is this.

I might want to use https://github.com/fsspec/gcsfs/blob/378cae6f56f3dbca1b909d58f21a8d61ed5b0a45/gcsfs/core.py#L321-L323 for the domain URL if I'll prepare a PR to the `gcsfs`.